# ALGORITMO 3
Este algoritmo calcula el n-ésimo número en la secuencia de Fibonacci de manera iterativa. Para n = 11, debemos encontrar el valor de y al final de la ejecución.

- Args:
        n (int): Posición en la secuencia de Fibonacci (n >= 0)
        
- Returns:
        int: El valor del n-ésimo número de Fibonacci

In [ ]:
# Versión con impresión de pasos para entender el proceso
def fibonacci_detallado(n):
    if n == 0:
        return 0
    
    x, y = 0, 1
    print(f"Paso inicial: x = {x}, y = {y} (F(0) y F(1))")
    
    for i in range(1, n):
        z = x + y
        print(f"Iteración {i}: x = {x}, y = {y} → z = {z} (F({i+1}))")
        x, y = y, z
        
    return y

print("\nProceso para n = 11:")
fibonacci_detallado(11)


Proceso detallado para n = 11:
Paso inicial: x = 0, y = 1 (F(0) y F(1))
Iteración 1: x = 0, y = 1 → z = 1 (F(2))
Iteración 2: x = 1, y = 1 → z = 2 (F(3))
Iteración 3: x = 1, y = 2 → z = 3 (F(4))
Iteración 4: x = 2, y = 3 → z = 5 (F(5))
Iteración 5: x = 3, y = 5 → z = 8 (F(6))
Iteración 6: x = 5, y = 8 → z = 13 (F(7))
Iteración 7: x = 8, y = 13 → z = 21 (F(8))
Iteración 8: x = 13, y = 21 → z = 34 (F(9))
Iteración 9: x = 21, y = 34 → z = 55 (F(10))
Iteración 10: x = 34, y = 55 → z = 89 (F(11))


89

# Optimización del Algoritmo Fibonacci

#### Versión original lenta

In [3]:
from functools import lru_cache
import time

def fib_slow(n):
    if n == 0: return 0
    x, y = 0, 1
    for _ in range(n):
        x, y = y, x + y
    return x

#### Versión optimizada (≈1000x más rápida)

In [4]:
def fib_fast(n, _cache={0: 0, 1: 1}):
    if n not in _cache:
        a, b = 0, 1
        for _ in range(n):
            a, b = b, a + b
        _cache[n] = a
    return _cache[n]

In [5]:
# Prueba de velocidad
n_values = range(5_000)

# Medición versión lenta
start = time.time()
[fib_slow(x) for x in n_values]
slow_time = time.time() - start

# Medición versión rápida
start = time.time()
[fib_fast(x) for x in n_values]
fast_time = time.time() - start

print(f"Tiempo original: {slow_time:.4f} segundos")
print(f"Tiempo optimizado: {fast_time:.4f} segundos")
print(f"Factor de mejora: {slow_time/fast_time:.0f}x más rápido")

Tiempo original: 2.0140 segundos
Tiempo optimizado: 1.8344 segundos
Factor de mejora: 1x más rápido


# Participación en clase
### P04: Usando redondeo de 3 cifras usar función FL_rounding
- Calcula en qué iteración la razón de Fibonacci y_{i+1}/y_i
    converge a la razón áurea con error relativo < tolerancia.

In [2]:
import math

def golden_ratio_error(max_iter=1000, tolerance=1e-5):

    # Valor exacto de la razón áurea
    gr = (1 + math.sqrt(5)) / 2
    
    # Inicialización de Fibonacci
    x, y = 0, 1
    iteration = 0
    
    for i in range(1, max_iter + 1):
        x, y = y, x + y
        
        if x == 0:
            continue  # Evitar división por cero
        
        # Calcular razón y error
        ratio = y /x
        error = abs((gr - ratio) / gr)
        
        # Redondear a 3 cifras como se pide
        ratio_rounded = round(ratio, 3)
        error_rounded = round(error, 3)
        
        print(f"Iteración {i}: Razón = {ratio_rounded}, Error = {error_rounded}")
        
        if error < tolerance:
            iteration = i
            break
    
    return iteration, error_rounded, ratio_rounded

# Calcular
gr = (1 + math.sqrt(5)) / 2
print(f"Razón áurea exacta: {gr:.5f}\n")

iteracion, error, razon = golden_ratio_error()
print(f"\nConvergencia en iteración {iteracion}")
print(f"Razón aproximada: {razon}")
print(f"Error relativo: {error} < 1e-5")

Razón áurea exacta: 1.61803

Iteración 1: Razón = 1.0, Error = 0.382
Iteración 2: Razón = 2.0, Error = 0.236
Iteración 3: Razón = 1.5, Error = 0.073
Iteración 4: Razón = 1.667, Error = 0.03
Iteración 5: Razón = 1.6, Error = 0.011
Iteración 6: Razón = 1.625, Error = 0.004
Iteración 7: Razón = 1.615, Error = 0.002
Iteración 8: Razón = 1.619, Error = 0.001
Iteración 9: Razón = 1.618, Error = 0.0
Iteración 10: Razón = 1.618, Error = 0.0
Iteración 11: Razón = 1.618, Error = 0.0
Iteración 12: Razón = 1.618, Error = 0.0
Iteración 13: Razón = 1.618, Error = 0.0

Convergencia en iteración 13
Razón aproximada: 1.618
Error relativo: 0.0 < 1e-5


# Ejercicio

### ¿Cuál es el resultado del algoritmo 03 cuando n = 11?

$z = x + y$

$x = y$

$y = z$

Si n es igual a 11 entonces se calcula hasta n - 1 = 10

- Iteración 1: $z = 0 + 1$, $x = 1$, $y = 1$
- Iteración 2: $z = 1 + 1$, $x = 1$, $y = 2$
- Iteración 3: $z = 1 + 2$, $x = 2$, $y = 3$
- Iteración 4: $z = 2 + 3$, $x = 3$, $y = 5$
- Iteración 5: $z = 3 + 5$, $x = 5$, $y = 8$
- Iteración 6: $z = 5 + 8$, $x = 8$, $y = 13$
- Iteración 7: $z = 8 + 13$, $x = 13$, $y = 21$
- Iteración 8: $z = 13 + 21$, $x = 21$, $y = 34$
- Iteración 9: $z = 21 + 34$, $x = 34$, $y = 55$
- Iteración 10: $z = 34 + 55$, $x = 55$, $y = 89$

Y el resultado de $y$ es igual a 89.


### Usando el algoritmo 03 y la aritmética de redondeo con 3 cifras determine la iteración desde la cual el error relativo  de $\frac{y_i + 1}{y_i}(i > 0)$ con respecto a $\frac{1 + \sqrt{5}}{2}$ esta dentro de $10^{-5}$

$gr =\frac{1 + \sqrt{5}}{2} = 1.618033989$

$error = \left|\frac{gr - \frac{y_i + 1}{y_i}}{gr}\right|$, $i > 0$

$error < 10^{-5}$, $i = ?$

- < 10

    $error = 2.100620267 \times 10^{-5} > 10^{-5}$

En las siguientes iteraciones donde i es igual a 100 y a 500 el gr al redondear a 3 cifras siempre sera 1.618 por lo que el error nunca sera menor a $10^{-5}$, entonces la opcion es **ninguna de las anteriores**.